### DISCUSSION TITLE
Would it be possible to output the "gaps" as a separate mask or fill them with a solid color instead of the current tiling modes?

### DISCUSSION THREAD

**Discussion by ProtoBelisarius** [BrokenSource/DepthFlow#58] [Ideas]
I have little knowledge on how shaders work, but I would like to use LaMa to inpaint the gaps instead of the current solutions with repeat/mirror.

So I was wondering if its feasible to output the frame with whatever tiling mode selected and a black/white mask depicting the gap.

Or directly fill the gaps in the frame with bright green or something instead of a tiling mode.

**Comment by Tremeschin** [comment]
Excellent question! 🙂

I've tried this in the past and couldn't get it just right, but also didn't try enough haha

I know a way to quantify the _gaps_, which brings me back to good old calculus 3 days. We need to compute the [gradient](https://en.wikipedia.org/wiki/Gradient) of each surface point (a normal vector with extra info), the more this vector is _tilted_ to the sides, the more we're on a _cliff_ in the scene

Calculating the _tilt_ is easy, it's the [angle between](https://en.wikipedia.org/wiki/Dot_product#Geometric_definition) the gradient and the scene's plane, but defining what are _gaps_ based on the _cliffs_ isn't inherently mathematical and subjective. We'll get a value between 0 and 1 (after normalization) telling how _steep_ a given pixel takes place on the scene's surface, which has zero knowledge of the surroundings and could give false positives

I'll give this a second chance in time, as I probably messed up the gradient calculation before

**Reply by Tremeschin** [reply]
This seems quite accurate (?), would you say it works for the use case of inpainting?

https://github.com/user-attachments/assets/0c1c93c5-db45-4ca7-bbc2-6ae243555815

I don't think I can do much better than this math-wise with my current shader experience, the artifacts are mostly due the relatively low resolution depth maps, and edges are.. complicated 😅

**Reply by KazusaKitakawa** [reply]
Hello! Thank you for your great contribution to this project. I'm also very interested in extracting masks and performing inpainting-based restoration. I would like to ask how the mask detection is implemented in your work. Is there any related code available for reference? @Tremeschin

**Reply by Tremeschin** [reply]
Surely have @KazusaKitakawa, it's a heuristic utilizing some calculus/geometry knowledge within the raymarching [(1)](https://en.wikipedia.org/wiki/Ray_marching) [(2)](https://www.youtube.com/watch?v=khblXafu7iA) model of the project. I basically keep track of the previous and current depthmap heights to estimate a [derivative](https://github.com/BrokenSource/DepthFlow/blob/ab63fb37721a7329a966072f49e27a69dd1bc258/depthflow/resources/shaders/depthflow.glsl#L103) value relative to the step size `quality` (which tells me how fast the surface height is changing); then use [finite differences](https://computergraphics.stackexchange.com/a/8097) to estimate a [normal vector](https://en.wikipedia.org/wiki/Normal_(geometry)) next in [the code](https://github.com/BrokenSource/DepthFlow/blob/ab63fb37721a7329a966072f49e27a69dd1bc258/depthflow/resources/shaders/depthflow.glsl#L111) to tell me the direction the change is happening

Combining the two, the further away the normal's angle is from the screen plane (I use Z=Forward), combined with how fast it changes, roughly gives me a _"global"_ perceptual [steepness](https://github.com/BrokenSource/DepthFlow/blob/ab63fb37721a7329a966072f49e27a69dd1bc258/depthflow/resources/shaders/depthflow.glsl#L118) value that depends on the surface max height

<sup><b>Note:</b> The projections are extremely [non-linear](https://github.com/BrokenSource/DepthFlow/blob/ab63fb37721a7329a966072f49e27a69dd1bc258/depthflow/resources/shaders/depthflow.glsl#L39-L61), a step can't have knowledge of the neighbors distortions to use a simpler approach</sup>

Then just about [checking for some](https://github.com/BrokenSource/DepthFlow/blob/ab63fb37721a7329a966072f49e27a69dd1bc258/depthflow/resources/shaders/depthflow.glsl#L161-L168) threshold and overriding the color as needed. No clue what the units are here!

**Reply by KazusaKitakawa** [reply]
Thanks a lot for your detailed reply! I’ll go through that part of the code. Great work!
